### ISSUE TITLE
Juvix needs a universe hierarchy to avoid the type-in-type contradiction

### ISSUE THREAD

**Issue by rokopt** [anoma/juvix#934]
Juvix's type system as currently implemented has a type of all types, which leads to inconsistency in Martin-Löf-style type theories (see for example [The Paradox of Trees in Type Theory](https://www.semanticscholar.org/paper/The-paradox-of-trees-in-type-theory-Coquand/f2f330b27f1d7ca99c2550f96581a4400c209ef8?p2df)).  We will have to implement type universes to avoid this.

@jonaprieto reminded me of this in the context of user-defined datatypes (#910 ).  The two are not directly related; this doesn't block #910.  I _do_ wonder whether we should treat it as a blocker for Stygarctus:  our general principle at least for datatypes is that we should restrict the release targets so as to produce only false negatives when typechecking, not false positives (i.e., we'll reject some things that will eventually typecheck, but we shouldn't accept anything that will eventually _not_ typecheck), but I'm not sure whether that principle includes type-in-type.

@cwgoes (among others) might have input on that.  @andy-morris might know whether universes would be sufficiently straightforward to target for Stygarctus or not.

**Comment by cwgoes** [comment]
Yes; we should probably do something like what Idris does (IIRC): figure out the directed graph (which types must be higher than which other types), find a satisfactory assignment of levels, and then typecheck ([ref](https://docs.idris-lang.org/en/latest/faq/faq.html#does-idris-have-universe-polymorphism-what-is-the-type-of-type)).

**Comment by jonaprieto** [comment]
Towards supporting cumulativity for universe polymorphism. We may want to consider adapting the algorithm described in [this repo](https://github.com/dybjer/universe-polymorphism).  The issue essentially is to solve a system of constraints in a semi-succ lattice formed by the universe levels.
